test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/resume/resume001.java

```diff
@@ -353,6 +352,13 @@ private int runThis (String argv[], PrintStream out) {
                     expresult = returnCode1;
                     break label1;
                 }
+
+                // We need to resume the main thread because thread2 might be blocked on it,
```

Comment:
This does not look correct to me.
This is the last test scenario - thread2.resume should resumes the thread while vm is suspended.
thread2 should not be blocked on main thread.
Looking at the debuggee I suppose the blocking is possible during logging. I'd suggest to update the debugee and remove any logging between breakpoints 2 and 3

Reply:
> This does not look correct to me.
> This is the last test scenario - thread2.resume should resumes the thread while vm is suspended.
> thread2 should not be blocked on main thread.
> Looking at the debuggee I suppose the blocking is possible during logging. I'd suggest to update the debugee and remove any logging between breakpoints 2 and 3

It looks like the debuggee gets as far as the following:

```
    public void runt2() {
        log("method 'runt2' enter 1");
        i1++;
        i2--;
        log("method 'run2t' exit");
        return;
    }

```

It prints the first log and hits a breakpoint setup on the 2nd line. The debugger resumes thread2 after this, but we never see the 2nd log. Whenever we see this failure, the following logs from the mainThread are always delayed (by a lot):

```
debugee.stderr> **> mainThread: mainThread is out of: synchronized (lockingObject) {
debugee.stderr> **> mainThread: waiting for an instruction from the debugger ...
```

I think this delay is resulting in the the mainThread being in the middle of doing one of these logs when the vm.suspend() is done. This leaves mainThread suspended while holding a lock needed for doing logging (logging is just a simple System.err.prinln()). I'm trying to prove this by getting a debuggee thread dump when getting the 3rd Breakpoint event times out, but for some reason once I added this code I could no longer reproduce the problem (still trying though).

I don't like the idea of avoiding this issue by getting rid of certain problematic logging. It seems error prone. Someone could add some new logging in the future. I'll see if there is a better solution than the vm.resume(). Perhaps I could just resume mainThread. However, I think with virtual threads I/O can be dependent on other threads like an "unparker" thread.

Another solution might be to have the debugger and debuggee do an additional handshake so we can guarantee that mainThread is done with these two log statements. Currently when we get to the 2nd log statement, that just means the debuggee is waiting for a "quit" command from the debugger. We could at this point have the debuggee send a command to the debugger, and have the debugger wait for this command before it does the vm.suspend().

Reply:
I was finally able to reproduce the issue with the stack dumping support in place. mainThread is in the middle of printing the 1st of the two logs mention above. mainThead is suspended and is holding a println related lock, and thread2 is stuck on the 2nd log call in runt2 awaiting the same lock.

Reply:
I was able to add synchronization between the debugger and debuggee to fix this issue, but I don't really like the solution. It just adds more complexity and makes the test even harder to follow. What I'd like to do is just put a short sleep in before the vm.suspend(). Let me know if you think this is ok and I'll update the PR with the changes.

Reply:
Thank you for the confirming the reason of the timeout.

To be more clear about my point:
The test has 3 scenarios (see the test description):
ThreadReference.resume() resumes the thread suspended with:
 *   - with thread2.suspend()                                           <BR>
 *   - at a breakpoint                                                  <BR>
 *   - with VirtualMachine.suspend()                                    <BR>

So for 3rd scenario we should not call vm.resume() (as it converts 3rd scenario to 1st scenario).
The test can be fixed by different ways, to me remove logging between breakpoint2 and breakpoint3 is the simplest way.
Note that breakpoint2 is "runt2(), line 2" and breakpoint3 is "runt1(), line 7", there are 2 log statements. We can move breakpoint 3 to "runt2(), line 3" (I don't see much sense to have breakpoint 3 so far from breakpoint2 - we just need to ensure the thread was resumed )

Reply:
Removing log() statements to fix the problem can be risky because someone could re-add them in the future. What about my idea of doing a short sleep before the vm.suspend() to make sure the main thread has advanced to the pipe.readln(), and won't be doing any more log calls until it gets the next command from the debugger (which should be "quit").

Reply:
Adding a `sleep()` between two statements does not scale when the test in
question is under different loads or run with different options. All it will do
is make a hang more rare (and frustrating to analyze).

We do use short sleeps when we are in a while-loop checking on a condition
of some sort that indicates that the "thing" for which we are waiting to happen
has happened.

Reply:
I believe we've done quite a few short sleeps like this in the past. Scaling is not really an issue. It should only require at most a few ms, even with -Xcomp, so we wait 1000ms and then never have to think about the timing again. This test is ugly. Sometimes you have to fight ugly with ugly.

Reply:
> mainThead is suspended and is holding a println related lock, and thread2 is stuck on the 2nd log call in runt2 awaiting the same lock.

The classic example of why suspension is fraught with peril - the target must be guaranteed not to be holding any resource needed by the suspender. I think removing the logging may be the best bet here - with comments in the code to ensure someone does not add it back. Or else use a more primitive (native?) mechanism to do the logging, not  System.out.println().

Reply:
Which logging should be removed? Alex suggest the logging between breakpoints 2 and 3, but even that is not enough. There is logging after breakpoint 3, and that happens before the vm.resume() is done. I'm not saying this can't be done right, but I think pruning out logging like this in order to fix the problem not only removes from the test of some valuable logging, but still leaves us open to this type of issue.

I think the safer thing to do is to make sure mainThread is no longer logging (or will attempt to log) when the vmsuspend is done. This could be done by pruning some of its logging, but that has the same negatives as removing some thread2 logging. My sleep suggestion is by far the simplest. The "purist" fix would probably be the checkpoint approach (don't do the vm.suspend until mainThread has reached a stable point). That ended up getting a bit uglier than I had hoped, but I can finish up the work so you can have a look at it if you'd like.

Reply:
Sorry I'm unclear on the different threads involved here. IIUC the vm.suspend comes from the debugger, and the mainthread and thread-2 are both threads in the debuggee, being suspended at different times?

Reply:
Yes. thread2 is suspended via breakpoint (multiple times). mainThread is suspended by the one place in the test that does a vm.suspend(), which is near the end of the test. This is the problematic suspend because sometimes it is done while mainThread is in the middle of a println. A bit later thread2 is resumed and ends up blocking on a println due to mainThread holding the needed lock.

Reply:
I've updated the implementation so now it does a sync point after mainThread is done with printlns.

Reply:
> Which logging should be removed? Alex suggest the logging between breakpoints 2 and 3, but even that is not enough. There is logging after breakpoint 3, and that happens before the vm.resume() is done.

After breakpoint 3 debugger calls thread2.resume() and vm.resume(), so there is no issue there (all threads are resumed).